azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -48,3 +63,331 @@ def _execute_op(self, op):
+                    retry_interval = (
+                        int(provisioning_op.retry_after, 10)
+                        if provisioning_op.retry_after is not None
+                        else constant.DEFAULT_POLLING_INTERVAL
+                    )
+                    decoded_response = self._decode_response(provisioning_op)
+                    operation_id = self._get_operation_id(decoded_response)
+
+                    # retry after scenario
+                    if success_status_code >= 429:
+
+                        self_weakref = weakref.ref(self)
+
+                        @pipeline_thread.invoke_on_pipeline_thread_nowait
+                        def do_retry_after():
+                            this = self_weakref()
+                            logger.info(
+                                "{stage_name}({op_name}): retrying".format(
+                                    stage_name=this.name, op_name=op.name
+                                )
+                            )
+                            op.retry_after_timer.cancel()
+                            op.completed = False
+                            this._execute_op(op)
+
+                        logger.warning(
+                            "{stage_name}({op_name}): Op needs retry with interval {interval} because of {error}.  Setting timer.".format(
+                                stage_name=self.name,
+                                op_name=op.name,
+                                interval=retry_interval,
+                                error=error,
+                            )
+                        )
+
+                        op.retry_after_timer = Timer(retry_interval, do_retry_after)
+                        op.retry_after_timer.start()
+
+                    # Service success scenario
+                    else:
+                        registration_status = self._get_registration_status(decoded_response)
+                        if registration_status == "assigned" or registration_status == "failed":
+                            # process complete response here
+                            complete_registration_result = self._form_complete_result(
+                                request_id=provisioning_op.request_id,
+                                operation_id=operation_id,
+                                decoded_response=decoded_response,
+                                status=registration_status,
+                            )
+                            op.registration_result = complete_registration_result
+                            self.complete_op(op, error=error)
+                        elif registration_status == "assigning":
+                            self.send_op_down(
+                                op=pipeline_ops_provisioning.SendQueryRequestOperation(
+                                    request_payload=" ",
+                                    operation_id=operation_id,
+                                    callback=op.callback,
```

Comment:
>callback=op.callback, [](start = 36, length = 21)

This concerns me.  Moving a callback from one op into another op will cause problems, though I'm not 100% sure what will happen.  I do know that the original SendRegistrationRequestOperation will probably become "orphaned" at this point.  When the SendQueryRequestOperation completes, it will call the (original) SendRegistrationRequestOperation callback with the (new) SendQueryRequestOperation object as the first parameter.  This, in effect, tells the caller that some new operation (which they know nothing about) just completed, and that is nonsensical.  

This is a scenario where the SendQueryReqestOperation is almost a worker_op for SendRegistrationRequestOperation, except, when the SendQueryReqestOperation completes, it needs to copy the registration result from the (new) SendQueryReqestOperation object into the (original) SendRegistrationRequestOperation object.  i think you need to write a callback for the (new) SendQueryReqestOperation which copies the result into the (original0 SendQueryRequestOperation and then completes that op.

Reply:
that was the reason I did worker op initially and later decided against it. Also I felt that the worker ops are converted op from the original op whereas the query was a separate op altogether.

---
In reply to: [345915183](https://github.com/Azure/azure-iot-sdk-python/pull/391#discussion_r345915183) [](ancestors = 345915183)